Getting Funqy Ready for Release (#9344)
quarkusio/quarkus: docs/src/main/asciidoc/funqy-amazon-lambda.adoc

Use the `invoke` command to invoke your function.

[source, subs=attributes+]
----
sh target/manage.sh invoke
----

The example function takes input passed in via the `--payload` switch which points to a json file
in the root directory of the project.

The function can also be invoked locally with the SAM CLI like this:

[source]
----
sam local invoke --template target/sam.jvm.yaml --event payload.json
----

If you are working with your native image build, simply replace the template name with the native version:

[source]
----
sam local invoke --template target/sam.native.yaml --event payload.json
----

== Update the function

You can update the Java code as you see fit.  Once you've rebuilt, you can redeploy your function by executing the
`update` command.

[source, subs=attributes+]
----
sh target/manage.sh update
----

== Deploy to AWS Lambda Custom (native) Runtime

If you want a lower memory footprint and faster initialization times for your Funqy function, you can compile your Java
code to a native executable.  Just make sure to rebuild your project with the `-Pnative` switch.

For Linux hosts execute:

[source, subs=attributes+]
----
mvn package -Pnative
----

> NOTE: If you are building on a non-Linux system, you will need to also pass in a property instructing quarkus to use a docker build as Amazon

Reviewer: ```suggestion
NOTE: If you are building on a non-Linux system, you will need to also pass in a property instructing Quarkus to use a docker build as Amazon
```

[source, subs=attributes+]
----
sh target/manage.sh native create
----

As above, commands can be stacked.  The only requirement is that `native` be the first parameter should you wish
to work with native image builds.  The script will take care of the rest of the details necessary to manage your native
image function deployments.

Examine the output of the `manage.sh` script if you want to learn what aws commands are executed
to create, delete, and update your functions.

One thing to note about the create command for native is that the `aws lambda create-function`
call must set a specific environment variable:

[source, subs=attributes+]
----
--environment 'Variables={DISABLE_SIGNAL_HANDLERS=true}'
----

== Examine the POM

There is nothing special about the POM other than the inclusion of the `quarkus-funqy-amazon-lambda` and `quarkus-test-amazon-lambda` extensions
as a dependencies.  The extension automatically generates everything you might need for your lambda deployment.

== Integration Testing

Funqy Amazon Lambda support leverages the Quarkus AWS Lambda test framework so that you can unit tests your Funqy functions.
This is true for both JVM and native modes.

> This test frameowrk provides similar functionality to the SAM CLI, without the overhead of Docker.

Reviewer: ```suggestion
This test framework provides similar functionality to the SAM CLI, without the overhead of Docker.
```

        Assertions.assertEquals("Hello Bill", out.getMessage());
    }
}
----

== Testing with the SAM CLI

The https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html[AWS SAM CLI]
allows you to run your functions locally on your laptop in a simulated Lambda environment.  This requires
https://www.docker.com/products/docker-desktop[docker] to be installed.  This is an optional approach should you choose
to take advantage of it.  Otherwise, the Quarkus JUnit integration should be sufficient for most of your needs.

A starter template has been generated for both JVM and native execution modes.

Run the following SAM CLI command to locally test your function, passing the appropriate SAM `template`.
The `event` parameter takes any JSON file, in this case the sample `payload.json`.

[source]
----
sam local invoke --template target/sam.jvm.yaml --event payload.json
----

The native image can also be locally tested using the `sam.native.yaml` template:

[source]
----
sam local invoke --template target/sam.native.yaml --event payload.json
----

== Modifying `function.zip`

> The are times where you may have to add some additions to the `function.zip` lambda deployment that is generated

Reviewer 2: ```suggestion
The are times where you may have to add some addition entries to the `function.zip` lambda deployment that is generated
```

by the build.  To do this create a `zip.jvm` or `zip.native` directory within `src/main`.
Create `zip.jvm/` if you are doing a pure Java.  `zip.native/` if you are doing a native deployment.

Any you files and directories you create under your zip directory will be included within `function.zip`

== Custom `bootstrap` script

> There are times you may want to set a specific system properties or other arguments when lambda invokes

Reviewer 2: ```suggestion
There are times you may want to set specific system properties or other arguments when lambda invokes
```
